feat: use p6spy's log filtering settings when tracing

high-stakes opened the pull request:
p6spy provides sql query filtering options which were not used during span
creation, added support for them.

p6spy tends to produce enormous amount of logs due to including all queries, sending this huge amount of spans is usually counterproductive ,adds noise to the analysis and increases costs

jorgheymans commented:
:+1:

perhaps missing is a small note in the readme that this is now supported ?

adriancole commented on instrumentation/p6spy/src/main/java/brave/p6spy/TracingP6SpyOptionsMBean.java:
```diff
@@ -0,0 +1,11 @@
+import com.p6spy.engine.logging.P6LogLoadableOptions;
+
+public interface TracingP6SpyOptionsMBean {
+    P6LogLoadableOptions getLogOptions();
```
is this a stable api?

high-stakes replied:
p6spy requires all options classes to have an mbean interface, previously this was implicitly already added because P6SpyOptions was extended (I think incorrectly because none of its options were used) and it already exposed an mbean interface

adriancole replied:
I mean to as if this P6LogLoadableOptions has been around for a while.. ex what version added etc. If we add this we could break others. Not all libraries track compatibility.

high-stakes replied:
hasn't been broken since the initial github release in 2013, 2.0.0

adriancole commented on instrumentation/p6spy/src/test/java/brave/p6spy/TracingJdbcEventListenerTest.java:
```diff
@@ -10,36 +10,48 @@
-import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.verifyNoMoreInteractions;
-import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.*;
```
please use square formatter it will protect from accidental wildcards etc

adriancole commented:
ps this looks like 2 changes JMX and also the logging thing

high-stakes commented on instrumentation/p6spy/src/main/java/brave/p6spy/TracingP6SpyOptions.java:
```diff
@@ -1,32 +1,61 @@
 import java.util.Map;
 
-final class TracingP6SpyOptions extends P6SpyOptions {
+public class TracingP6SpyOptions extends StandardMBean implements P6LoadableOptions, TracingP6SpyOptionsMBean {
```
in line with the other options classes of p6spy

adriancole replied:
nothing should be public unless it is required to. each public type adds a possibility for us to break api. there's a place in effective java about this, but it is something followed (or accidentally not followed) in this codebase even if p6spy doesn't follow the same coding rules we do. In this repo, we prefer our coding rules.

high-stakes commented on instrumentation/p6spy/src/main/java/brave/p6spy/TracingP6SpyOptions.java:
```diff
@@ -1,32 +1,61 @@
-  String remoteServiceName() {
+  @Override
+  public Map<String, String> getDefaults() {
+    Map<String, String> allDefaults = new HashMap();
```
any specific reason or just preference?

high-stakes commented on instrumentation/p6spy/src/main/java/brave/p6spy/TracingP6SpyOptionsMBean.java:
```diff
@@ -0,0 +1,11 @@
+
+import com.p6spy.engine.logging.P6LogLoadableOptions;
+
+public interface TracingP6SpyOptionsMBean {
```
not sure, I followed the p6spy convention

adriancole replied:
either remove this or add a unit test to prove it works. right now it is a distracting file

high-stakes commented:
> another option is a finished span handler that drops these. such wouldnt be as subject to api drift. you wouldnt need to change the code here at all or conflate logging with span recording..

Sounds cool, maybe I could add both? I will look into this on the weekend

adriancole commented:
FYI if adding a change to directly implement JMX, such a PR will need a unit test to prove what we support changing can be done via JMX.

regardless, thanks for the help!

high-stakes commented:
> FYI if adding a change to directly implement JMX, such a PR will need a unit test to prove what we support changing can be done via JMX.
> 
> regardless, thanks for the help!

high-stakes commented:
> FYI if adding a change to directly implement JMX, such a PR will need a unit test to prove what we support changing can be done via JMX.
> 
> regardless, thanks for the help!

I mean the previous usage was incorrect (it would do nothing best case, worst case replace the officially registered MBean for P6SpyOptions which is always implicitly loaded), I just removed it to avoid bugs. Is it enough if i add an empty interface and not expose any methods?

adriancole commented on instrumentation/p6spy/src/main/java/brave/p6spy/TracingP6SpyOptions.java:
```diff
@@ -1,32 +1,58 @@
 
-  TracingP6SpyOptions(P6OptionsRepository optionsRepository) {
-    super(optionsRepository);
+  public TracingP6SpyOptions(P6OptionsRepository optionsRepository) {
```
revert public here. note: please use the format on all files changed as I notice some missing or too many spaces around. if you look at the diff you can notice.

adriancole commented on instrumentation/p6spy/src/main/java/brave/p6spy/TracingP6SpyOptions.java:
```diff
@@ -1,32 +1,58 @@
+  @Override
+  public Map<String, String> getDefaults() {
+    Map<String, String> allDefaults = new LinkedHashMap();
+    allDefaults.putAll(defaults);
```
inline`defaults` vs a static initializes for the hashmap which only has one attribute.

adriancole commented on instrumentation/p6spy/src/main/java/brave/p6spy/TracingP6SpyOptions.java:
```diff
@@ -1,32 +1,58 @@
+    return allDefaults;
+  }
+
+  public P6LogLoadableOptions getLogOptions() {
```
PS If you want to add an interface, just make the interface itself package private. Then, you can add `@Override` tag for this and below. As long as types are package private, we're good. 

f we are broken on JMX and don't know it because there is no test, a good fix would be to add a test vs us guessing if it is broke or not. I suspect there's code out there somewhere that has a unit test for JMX registration even if we don't have any in this repo. I'd consider this part optional for this PR even if we are broken as ensuring JMX works can be separate

high-stakes commented:
> as p6spy can work from properties (it has so far) please either remove all JMX related changes or add a unit test to show that they work.
> 
> separately, I also left comments with some style-related fuzz. The core change here about log options is done, it is just in the middle of a bunch of other things about JMX which are incomplete and distracting. If you can revert that into a different PR that has a unit test to show it works, we'll all save time.

Makes sense, added the changes and reverted the jmx part.

high-stakes commented:
Hi @adriancole , could you merge this? The changes discussed are added. Thanks

adriancole commented on instrumentation/p6spy/README.md:
```diff
@@ -42,3 +42,10 @@ jdbc:mysql://127.0.0.1:3306/mydatabase?zipkinServiceName=myServiceName
 
 The current tracing component is used at runtime. Until you have instantiated `brave.Tracing`, no traces will appear.
+
+### Log filtering
```
```suggestion
### Filtering spans
```

adriancole commented on instrumentation/p6spy/README.md:
```diff
@@ -42,3 +42,10 @@ jdbc:mysql://127.0.0.1:3306/mydatabase?zipkinServiceName=myServiceName
 The current tracing component is used at runtime. Until you have instantiated `brave.Tracing`, no traces will appear.
+
+### Log filtering
+
```
```suggestion

By default, all statements are recorded as client spans. You may wish to exclude statements like `set session` from tracing. This library reuses p6spy's log filtering for this purpose.

```

adriancole commented on instrumentation/p6spy/README.md:
```diff
@@ -42,3 +42,10 @@ jdbc:mysql://127.0.0.1:3306/mydatabase?zipkinServiceName=myServiceName
+### Log filtering
+
+The instrumentation also supports picking up p6spy's log filtering options from `spy.properties`, so you can filter 
+what type of statements to record.
```
```suggestion
what type of statements to record as spans.
```

high-stakes commented:
done!

adriancole commented:
many thanks @high-stakes and @jorgheymans for the review
